### PR TITLE
bug(python): fix path handling in windows

### LINKED ISSUE #703: Changes to dataset URI creation for LanceTable
This is a bit of a niche problem that would matter primarily for Windows users, but dataset URI creation within `lancedb.table.LanceTable._dataset_uri` right now makes use of `os.path.join` which, on a Windows machine, means that a connection URI joined with a table name inserts backslashes instead of forward slashes. Would it be possible to either have this section modified to account for different filesystems? Apologies if this is already solvable / fixable via some method that I'm not aware of

### PR DESCRIPTION
**Pull request by changhiskhan**
Use pathlib for local paths so that pathlib
can handle the correct separator on windows.

Closes #703

### CODE REVIEW

**Comment by sergiocorreia** [comment]
I also found this related path: https://github.com/lancedb/lancedb/issues/673#issuecomment-1850692449 might be worth it to deal with both in the same PR

File: python/lancedb/util.py [hunk 1]
```diff
@@ -84,6 +85,18 @@ def fs_from_uri(uri: str) -> Tuple[pa_fs.FileSystem, str]:
     return pa_fs.FileSystem.from_uri(uri)
 
 
+def join_uri(base: Union[str, pathlib.Path], *parts: str) -> str:
+    """
+    Join a URI with multiple parts, handling extra environment variables.
```
Comment (wjones127) [inline python/lancedb/util.py:90]: "extra environment variables"?

**Review by wjones127 — COMMENTED** [review]
Is there a test we can write for windows to validate this?

File: python/lancedb/util.py [hunk 2]
```diff
@@ -84,6 +85,18 @@ def fs_from_uri(uri: str) -> Tuple[pa_fs.FileSystem, str]:
...
+    """
+    if isinstance(base, pathlib.Path):
```
Comment (eddyxu) [inline python/lancedb/util.py:92]: Can we add tests?

**Comment by wjones127** [comment]
There seems to already be a test that fails on Windows due to path handling: https://github.com/lancedb/lancedb/actions/runs/7280056572/job/19837672999?pr=729

I have rebased your PR on my PR to add a Windows CI job.

File: python/lancedb/util.py (hunk 1 above)
Comment (changhiskhan) [reply python/lancedb/util.py:90]: oops that was generated by copilot and I forgot to change it >_<

**Comment by changhiskhan** [comment]
> I also found this related bug: [#673 (comment)](https://github.com/lancedb/lancedb/issues/673#issuecomment-1850692449) might be worth it to deal with both in the same PR

good catch. I added a condition so that if the uri scheme is one letter then just return the uri instead of the netloc + path or path. Lmk if this addresses the cases you've encountered on windows.

**Review by wjones127 — APPROVED** [review]

**Pull request merged**